src/test/java/org/jenkinsci/plugins/gitclient/GitAPITest.java

```diff
@@ -1489,6 +1488,51 @@ public void testLocalCheckoutConflict() throws Exception {
         assertEquals("old", FileUtils.readFileToString(workspace.file("foo"), "UTF-8"));
     }
 
+    @Test
+    public void testNoSubmodules() throws Exception {
+        workspace.touch(testGitDir, "committed-file", "committed-file content " + java.util.UUID.randomUUID().toString());
+        testGitClient.add("committed-file");
+        testGitClient.commit("commit1");
+        IGitAPI igit = (IGitAPI) testGitClient;
+        igit.submoduleClean(false);
+        igit.submoduleClean(true);
+        igit.submoduleUpdate().recursive(false).execute();
+        igit.submoduleUpdate().recursive(true).execute();
+        igit.submoduleSync();
+        assertTrue("committed-file missing at commit1", workspace.file("committed-file").exists());
+    }
+
+    @Test
+    public void testHasSubmodules() throws Exception {
+        workspace.launchCommand("git", "fetch", workspace.localMirror(), "tests/getSubmodules:t");
+        testGitClient.checkout().ref("t").execute();
+        assertTrue(testGitClient.hasGitModules());
+
+        workspace.launchCommand("git", "fetch", workspace.localMirror(), "master:t2");
+        testGitClient.checkout().ref("t2").execute();
+        assertFalse(testGitClient.hasGitModules());
+        assertFixSubmoduleUrlsThrows();
+    }
+
+    public void assertFixSubmoduleUrlsThrows() throws InterruptedException {
```

Comment:
If you're willing, I'm curious if this could be converted to use the JUnit 4 `ExpectedException` rather than try / catch / fail.  I'm also OK if you decide that it is not worth the effort to make that transition at this time.

Reply:
I will try for it. And let you know. I need to look into documentation of `ExpectedException` of JUnit 4.